docs/posts/pydata-performance/index.qmd

```diff
@@ -0,0 +1,696 @@
+
+```{python}
+#| echo: false
+#| output: asis
+with open("./step3.py") as f:
+    source = f.read()
+print(f"```python\n{source}\n```")
+exec(source)
+```
+
+1. I added a second sort key (`project_count`) for deterministic output.
+
+Let's run it to make sure the results are as expected:
+
+```{python}
+duckdb_results = %timeit -n1 -r1 -o expr.to_pandas()
+```
+
+It looks like the new query might be a bit slower even though we're ostensibly
+doing less computation. Since we're still pretty close to the original
+duration, let's keep going.
+
+```{python}
+#| echo: false
+#| output: asis
+gc.collect();
+```
+
+### Final pandas run with the new query
+
+Rewriting and running the pandas code we get:
+
+```{python}
+#| echo: false
+#| output: asis
+with open("./pandas_impl.py") as f:
+    print(f"```python\n{f.read()}\n```")
+```
+
+Running it we get:
+
+```{python}
+pandas_results = %timeit -n1 -r1 -o %run pandas_impl.py
+```
+
+```{python}
+#| echo: false
+#| output: asis
+del df
+gc.collect()
+del gc
+```
+
+::: {.callout-note}
+# Remember, this is the time it took pandas to run the query for a **single** file.
+DuckDB runs the query over the **entire** dataset about 4x faster than that!
+:::
+
+Let's try a tool that nominally scales to our problem: [Dask](https://dask.org).
+
+## Dask
+
+One really nice component of Dask is
+[`dask.dataframe`](https://docs.dask.org/en/stable/dataframe.html).
+
+Dask DataFrame implements a [good chunk of the pandas
+API](https://docs.dask.org/en/stable/dataframe.html#scope) and can be a drop-in
+replacement for pandas.
+
+I am happy that this turned out to be the case here.
+
+My first attempt was somewhat naive and was effectively a one line change
+from `import pandas as pd` to `import dask.dataframe as pd`.
+
+This worked and the workload completed. However, after talking to Dask
+expert and Ibis contributor [Naty Clementi](https://github.com/ncclementi) she
+suggested I try a few things:
+
+* Use [the distributed scheduler](https://distributed.dask.org/en/stable/).
+* Ensure that [`pyarrow` string arrays are
+  used](https://docs.dask.org/en/latest/configuration.html#dask) instead of
+  numpy object arrays.
+
+Both of those settings together shaved about 2x off the
+non-distributed-scheduler, non-pyarrow-string version while adding only
+a couple lines of code. Nice!
+
+Let's look at the Dask implementation:
+
+```{python}
+#| echo: false
+#| output: asis
+with open("./dask_impl.py") as f:
+    print(f"```python\n{f.read()}\n```")
+```
+
+Let's run the code:
+
+```{python}
+dask_results = %timeit -n1 -r1 -o %run dask_impl.py
+```
+
+That's a nice improvement over pandas: we finished the workload. Dask is still
+good ways behind DuckDB out of the box!
+
+## Takeaways
+
+**Ibis + DuckDB is the only system tested that handles this workload well out of the box**
+
+Pandas couldn't handle the workload due to memory constraints, and Dask
+required some fairly advanced knowledge to get maximum performance.
+
```

Comment:
I don't think that this is accurate. Arrow strings are on by default when you have a new Dask version and the distributed scheduler is in 10 minutes to Dask (https://docs.dask.org/en/stable/10-minutes-to-dask.html#scheduling)

Reply:
Happy to change it.

I think it's worth nothing that I am immediately presented with having to use `Client` and having to think about `fork` behavior making the out of the box experience a little less seamless than it perhaps could be.

Reply:
Yes 100% agree with you on this (I was just contesting the advanced part, because I think that's not correct)

Reply:
For example, the code I wrote originally didn't use the distributed scheduler and I couldn't port it without stick all the code under `if __name__ == "__main__":`

Reply:
Adjusted this bit, let me know what you think.